services/src/main/java/org/keycloak/services/DefaultKeycloakSession.java

```diff
@@ -169,9 +171,11 @@ public UserFederatedStorageProvider userFederatedStorage() {
     @Override
     @Deprecated
     public UserProvider userLocalStorage() {
-        // TODO: if we would call users() here, we could get the cache in legacy mode instead and would then loop
-        //   Also: this is still called in some places to avoid calling the federated storage
-        return getProvider(UserProvider.class);
+        if (log.isEnabled(Logger.Level.WARN)) {
+            // check if warning is enabled first before constructing the exception that is expensive to construct
+            log.warn("The semantics of this method have changed: If you require to talk to the local storage, consider calling LegacyDatastoreProvider#userLocalStorage() instead.", new RuntimeException());
```

Comment:
Please do not create the exception. Instead use `StackUtil.getShortStackTrace()` which is heap-saving: only single reusable object gets created.